examples_x035/usbpd_sink/usbpd.h

```diff
@@ -0,0 +1,921 @@
+		( ( pdo )->EPR_AVS.MaxVoltageIn100mV * 100 ), ( ( pdo )->EPR_AVS.PeakCurrent )
+
+typedef struct
+{
+	uint32_t MaxCurrent15To20VIn10mA : 10u;
+	uint32_t MaxCurrent9to15VIn10mA : 10u;
+	uint32_t Reserved_20_25bit : 6u; // shall be set to zero
+	USBPD_PeakCurrent_e PeakCurrent : 2u;
+	USBPD_AugmentedPDO_e AugmentedType : 2u; // shall be eUSBPD_APDO_SPR_AVS
+	USBPD_PowerDataObject_e PDOType : 2u; // shall be eUSBPD_PDO_AUGMENTED
+} USBPD_SPR_AVS_APDO_t;
+
+#define SPR_AVS_FMT                 \
+	"\nSPR AVS Supply:\n"           \
+	"\tMax Current 15-20V: %d mA\n" \
+	"\tMax Current 9-15V: %d mA\n"  \
+	"\tPeak Current: %d\n"
+
+#define SPR_AVS_FMT_ARGS( pdo )                                                                          \
+	( ( pdo )->SPR_AVS.MaxCurrent15To20VIn10mA * 10 ), ( ( pdo )->SPR_AVS.MaxCurrent9to15VIn10mA * 10 ), \
+		( ( pdo )->SPR_AVS.PeakCurrent )
+
+typedef union
+{
+	USBPD_PDOHeader_t Header;
+	USBPD_SourceFixedSupplyPDO_t FixedSupply;
+	USBPD_VariablePDO_t VariableSupply;
+	USBPD_BatteryPDO_t BatterySupply;
+	USBPD_SPR_PPS_APDO_t SPR_PPS;
+	USBPD_EPR_AVS_APDO_t EPR_AVS;
+	USBPD_SPR_AVS_APDO_t SPR_AVS;
+} USBPD_SourcePDO_t;
+
+static_assert( sizeof( USBPD_SourcePDO_t ) == sizeof( uint32_t ), "USBPD_SourcePDO_t size mismatch" );
+
+typedef enum // Do not PACK, messes up alignment
+{
+	eUSBPD_FAST_ROLE_SWAP_NOT_SUPPORTED = 0,
+	eUSBPD_FAST_ROLE_SWAP_DEFAULT = 1,
+	eUSBPD_FAST_ROLE_SWAP_1A5 = 2, // 1.5A @ 5V
+	eUSBPD_FAST_ROLE_SWAP_3A = 3, // 3A @ 5V
+} USBPD_FastRoleSwapRequiredCurrent_e;
+
+typedef struct
+{
+	uint32_t CurrentIn10mA : 10u;
+	uint32_t VoltageIn50mV : 10u;
+	uint32_t Reserved_20_22bit : 3u; // shall be set to zero
+	USBPD_FastRoleSwapRequiredCurrent_e FastRoleSwap : 2u;
+	uint32_t DualRoleData : 1u;
+	uint32_t USBComsCapable : 1u;
+	uint32_t UnconstrainedPower : 1u;
+	uint32_t HigherCapability : 1u;
+	uint32_t DualRolePower : 1u;
+	USBPD_PowerDataObject_e PDOType : 2u; // shall be eUSBPD_PDO_FIXED
+} USBPD_SinkFixedSupplyPDO_t;
+
+typedef union
+{
+	USBPD_PDOHeader_t Header;
+	USBPD_SinkFixedSupplyPDO_t FixedSupply;
+	USBPD_VariablePDO_t VariableSupply;
+	USBPD_BatteryPDO_t BatterySupply;
+	USBPD_SPR_PPS_APDO_t SPR_PPS;
+	USBPD_EPR_AVS_APDO_t EPR_AVS;
+	USBPD_SPR_AVS_APDO_t SPR_AVS;
+} USBPD_SinkPDO_t;
+
+static_assert( sizeof( USBPD_SourcePDO_t ) == sizeof( uint32_t ), "USBPD_SourcePDO_t size mismatch" );
+static_assert( sizeof( USBPD_SinkPDO_t ) == sizeof( uint32_t ), "USBPD_SinkPDO_t size mismatch" );
+
+typedef union
+{
+	USBPD_SourcePDO_t Source[7];
+	USBPD_SinkPDO_t Sink[7];
+} USBPD_SPR_CapabilitiesMessage_t;
+
+static_assert( sizeof( USBPD_SPR_CapabilitiesMessage_t ) == ( 7 * sizeof( uint32_t ) ),
+	"USBPD_SPR_CapabilitiesMessage_t size mismatch" );
+
+typedef struct
+{
+	uint32_t MaxCurrentIn10mA : 10u;
+	uint32_t OperatingCurrentIn10mA : 10u;
+	uint32_t Reserved_20_21bit : 2u; // shall be set to zero
+	uint32_t ERPCapable : 1u;
+	uint32_t UnchunkedExtendedMessage : 1u;
```

Comment:
I wanted to make sure - where did all these come from?  Are they completely your own work? Did you copy/paste from another source?

Reply:
All of these are written by me, by hand, from the PD spec. I have not even looked at any other USB PD implementations.